Assignment 1/src/main/java/com/cezaram28/Assignment1/service/VoteManagementService.java

```diff
@@ -28,12 +34,157 @@ public void remove(int id) {
     }
 
     @Transactional
-    public Optional<Vote> findByQuestion(int questionId, int userId) {
-        return repositoryFactory.createVoteRepository().findByQuestion(questionId, userId);
+    public Vote findByQuestion(int questionId, int userId) {
+        return repositoryFactory.createVoteRepository().findByQuestion(questionId, userId).orElseThrow(VoteNotFoundException::new);
     }
 
     @Transactional
-    public Optional<Vote> findByAnswer(int answerId, int userId) {
-        return repositoryFactory.createVoteRepository().findByAnswer(answerId, userId);
+    public Vote findByAnswer(int answerId, int userId) {
+        return repositoryFactory.createVoteRepository().findByAnswer(answerId, userId).orElseThrow(VoteNotFoundException::new);
+    }
+
+    @Transactional
+    public void upvoteQuestion(Question question, User user) {
+        if (question.getAuthor().getId() == user.getId()) {
+            throw new YourPostException();
+        } else {
+            Vote vote = new Vote(null, "up", question, null, user);
+            try {
+                Vote v = findByQuestion(question.getId(), user.getId());
+                if (v.getType().equals(vote.getType())) {
+                    throw new UpvotedException();
+                } else {
+                    // compute score by modifying parameters +7 question author, +2 question
+                    v.setType("up");
```

Comment:
In general, it would have been better to use an Enum instead of a simple string for the vote type.